### PR TITLE
feat: Escrow RPCs

### PR DESCRIPTION
**Pull request by theodorebugnet**
Adds RPCs for the escrow crate, to avoid functionality being re-implemented in the lib.

### CODE REVIEW

File: crates/escrow/src/lib.rs [hunk 1]
```diff
@@ -578,6 +578,13 @@ impl<T: Config> Pallet<T> {
         let last_point = <PointHistory<T>>::get(<Epoch<T>>::get());
         Self::supply_at(last_point, height)
     }
+
+    /// RPC only
+    pub fn total_locked() -> BalanceOf<T> {
```
Comment (theodorebugnet) [inline crates/escrow/src/lib.rs:583]: Incidentally, getting the actual total locked (staked) is something the lib also needs, but I don't see a way to get it without simply looping over all the locked stakes. Not sure if this is acceptable for an RPC call only (as it won't be used in any transactions) - or if it should just be left as lib functionality.

File: crates/escrow/rpc/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    Balance: Codec + MaybeDisplay + MaybeFromStr,
+    BlockNumber: Codec,
+    AccountId: Codec,
+{
+    #[method(name = "escrow_balanceAt")]
+    fn balance_at(
+        &self,
+        account_id: AccountId,
+        height: Option<BlockNumber>,
+        at: Option<BlockHash>,
```
Comment (gregdhill) [inline crates/escrow/rpc/src/lib.rs:32]: Having two block parameters is mildly confusing, I wonder if we can merge?

File: crates/escrow/rpc/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    Balance: Codec + MaybeDisplay + MaybeFromStr,
+{
+    fn balance_at(
+        &self,
+        account_id: AccountId,
+        height: Option<BlockNumber>,
+        at: Option<<Block as BlockT>::Hash>,
+    ) -> RpcResult<BalanceWrapper<Balance>> {
+        let api = self.client.runtime_api();
+        let at = BlockId::hash(at.unwrap_or_else(|| self.client.info().best_hash));
```
Comment (gregdhill) [inline crates/escrow/rpc/src/lib.rs:88]: If we do decide to merge the block parameters we could do something like this (roughly - needs error handling):
```suggestion
        let height = at.map_or_else(|| self.client.info().best_number, |hash| self.client.number(hash));
        let at = BlockId::hash(at.unwrap_or_else(|| self.client.info().best_hash));
```

File: crates/escrow/src/lib.rs [hunk 4]
```diff
@@ -578,6 +578,13 @@ impl<T: Config> Pallet<T> {
...
+        Locked::<T>::iter().fold(BalanceOf::<T>::default(), |sum, (_, locked)| {
+            sum.saturating_add(locked.amount)
+        })
+    }
```
Comment (gregdhill) [inline crates/escrow/src/lib.rs:587]: Not 100% sure about including this, would also ask for @sander2's opinion. There are however two alternatives:
- add a `TotalLocked` storage item
- move the lib call to squid (**recommended**)

**Review by gregdhill — COMMENTED** [review]
Only blocking comment is the inclusion of `total_locked()`, otherwise LGTM!

File: crates/escrow/src/lib.rs (hunk 1 above)
Comment (gregdhill) [reply crates/escrow/src/lib.rs:583]: The problem with this call is the potential for Denial of Service (DoS) but we anyway have looping operations in other RPC calls so I'm not sure. However, as mentioned in my other comment the best place for this functionality would be in squid.

File: crates/escrow/src/lib.rs (hunk 4 above)
Comment (sander2) [reply crates/escrow/src/lib.rs:587]: This is iterating over ~1400 items on Kintsugi so yea I'd agree that it's a good idea to use one of the alternatives suggested by Greg

**Review by sander2 — APPROVED** [review]

File: crates/escrow/rpc/src/lib.rs (hunk 3 above)
Comment (sander2) [reply crates/escrow/rpc/src/lib.rs:85]: Block-height can be a future block so we can't really merge

**Pull request merged**